irohad/consensus/yac/impl/yac_gate_impl.cpp

```diff
@@ -51,59 +51,72 @@ namespace iroha {
-                    return delayed_message;
-                  },
-                  rxcpp::identity_current_thread());
-              rxcpp::observable<GateObject> events =
-                  delayed_outcomes.flat_map([this](auto message) {
-                    return visit_in_place(message,
-                                          [this](const CommitMessage &msg) {
-                                            return this->handleCommit(msg);
-                                          },
-                                          [this](const RejectMessage &msg) {
-                                            return this->handleReject(msg);
-                                          },
-                                          [this](const FutureMessage &msg) {
-                                            return this->handleFuture(msg);
-                                          });
-                  });
-              rxcpp::connectable_observable<GateObject> published_events =
-                  events.publish();
-              rxcpp::observable<GateObject> published_ref_counted_events =
-                  published_events.ref_count();
-              return published_ref_counted_events;
-            }()),
             orderer_(std::move(orderer)),
             hash_provider_(std::move(hash_provider)),
             block_creator_(std::move(block_creator)),
             consensus_result_cache_(std::move(consensus_result_cache)),
-            hash_gate_(std::move(hash_gate)) {
-        block_creator_->onBlock().subscribe(
-            [this](const auto &event) { this->vote(event); });
+            hash_gate_(std::move(hash_gate)),
+            delay_func_(std::move(delay_func)) {}
+
+      void YacGateImpl::initialize() {
+        block_creator_subscription_ =
+            iroha::SubscriberCreator<bool, simulator::BlockCreatorEvent>::
+                template create<EventTypes::kOnBlockCreatorEvent,
+                                SubscriptionEngineHandlers::kYac>(
+                    [wptr(weak_from_this())](auto &, auto event) {
+                      if (auto ptr = wptr.lock())
+                        ptr->vote(event);
+                    });
+
+        delayed_outcome_subscription_ =
+            iroha::SubscriberCreator<bool, Answer>::template create<
+                EventTypes::kOnOutcomeDelayed,
+                SubscriptionEngineHandlers::kYac>([wptr(weak_from_this())](
+                                                      auto &, auto message) {
+              // check ptr ref remains 1
+              if (auto ptr = wptr.lock())
+                visit_in_place(
+                    message,
+                    [&](const CommitMessage &msg) { ptr->handleCommit(msg); },
+                    [&](const RejectMessage &msg) { ptr->handleReject(msg); },
+                    [&](const FutureMessage &msg) { ptr->handleFuture(msg); });
+            });
+
+        outcome_subscription_ =
```

Comment:
Language logic error. You create a Subscriber. But result variable is subscription. Please select what is result subscriber or subscription. subscription is result of subscriber's subscribe().
*Let's be clear in human language and code will also be more clear*